src/test/app/PayStrand_test.cpp

```diff
@@ -0,0 +1,1416 @@
+    jtx::Account const& dst,
+    Currency const& cur,
+    TrustFlag flag)
+{
+    if (auto sle = env.le(keylet::line(src, dst, cur)))
+    {
+        auto const useHigh = src.id() > dst.id();
+        return sle->isFlag(trustFlag(flag, useHigh));
+    }
+    Throw<std::runtime_error>("No line in getTrustFlag");
+    return false;  // silence warning
+}
+
+bool
+equal(std::unique_ptr<Step> const& s1, DirectStepInfo const& dsi)
+{
+    if (!s1)
+        return false;
+    return test::directStepEqual(*s1, dsi.src, dsi.dst, dsi.currency);
+}
+
+bool
+equal(std::unique_ptr<Step> const& s1, XRPEndpointStepInfo const& xrpsi)
+{
+    if (!s1)
+        return false;
+    return test::xrpEndpointStepEqual(*s1, xrpsi.acc);
+}
+
+bool
+equal(std::unique_ptr<Step> const& s1, ripple::Book const& bsi)
+{
+    if (!s1)
+        return false;
+    return bookStepEqual(*s1, bsi);
+}
+
+template <class Iter>
+bool
+strandEqualHelper(Iter i)
+{
+    // base case. all args processed and found equal.
+    return true;
+}
+
+template <class Iter, class StepInfo, class... Args>
+bool
+strandEqualHelper(Iter i, StepInfo&& si, Args&&... args)
+{
+    if (!equal(*i, std::forward<StepInfo>(si)))
+        return false;
+    return strandEqualHelper(++i, std::forward<Args>(args)...);
+}
+
+template <class... Args>
+bool
+equal(Strand const& strand, Args&&... args)
+{
+    if (strand.size() != sizeof...(Args))
+        return false;
+    if (strand.empty())
+        return true;
+    return strandEqualHelper(strand.begin(), std::forward<Args>(args)...);
+}
+
+STPathElement
+ape(AccountID const& a)
+{
+    return STPathElement(
+        STPathElement::typeAccount, a, xrpCurrency(), xrpAccount());
+};
+
+// Issue path element
+STPathElement
+ipe(Issue const& iss)
+{
+    return STPathElement(
+        STPathElement::typeCurrency | STPathElement::typeIssuer,
+        xrpAccount(),
+        iss.currency,
+        iss.account);
+};
+
+// Issuer path element
+STPathElement
+iape(AccountID const& account)
+{
+    return STPathElement(
+        STPathElement::typeIssuer, xrpAccount(), xrpCurrency(), account);
+};
+
+// Currency path element
+STPathElement
+cpe(Currency const& c)
+{
+    return STPathElement(
+        STPathElement::typeCurrency, xrpAccount(), c, xrpAccount());
+};
+
+// All path element
+STPathElement
+allpe(AccountID const& a, Issue const& iss)
+{
+    return STPathElement(
+        STPathElement::typeAccount | STPathElement::typeCurrency |
+            STPathElement::typeIssuer,
+        a,
+        iss.currency,
+        iss.account);
+};
+
+class ElementComboIter
+{
+    enum class SB /*state bit*/
+    { acc,
+      iss,
+      cur,
+      rootAcc,
+      rootIss,
+      xrp,
+      sameAccIss,
+      existingAcc,
+      existingCur,
+      existingIss,
+      prevAcc,
+      prevCur,
+      prevIss,
+      boundary,
+      last };
+
+    std::uint16_t state_ = 0;
+    static_assert(static_cast<size_t>(SB::last) <= sizeof(state_) * 8, "");
```

Comment:
Appveyor suggests Visual Studio can't handle this `static_assert`.  This might work for all our compilers...
```
static_assert (static_cast<size_t>(SB::last) <= sizeof(decltype (state_)) * 8, "");
```